src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/AnalyzerAssemblyService.cs

```diff
@@ -0,0 +1,213 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.ComponentModel.Composition;
+using System.Reflection;
+using System.Reflection.Metadata;
+using System.Reflection.PortableExecutable;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.Tree.Dependencies
+{
+    [Export(typeof(IAnalyzerAssemblyService))]
+    [AppliesTo(ProjectCapability.DependenciesTree)]
+    internal class AnalyzerAssemblyService : IAnalyzerAssemblyService
+    {
+        private object _guard = new object();
+        private ImmutableDictionary<string, bool> _map = ImmutableDictionary<string, bool>.Empty.WithComparers(StringComparers.Paths, EqualityComparer<bool>.Default);
+
+        public bool ContainsDiagnosticAnalyzers(string fullPathToAssembly)
+        {
+            if (_map.TryGetValue(fullPathToAssembly, out bool cachedValue))
+            {
+                return cachedValue;
+            }
+
+            var result = ContainsDiagnosticAnalyzersCore(fullPathToAssembly);
```

Comment:
Why is this outside of the lock? Do we want multiple threads computing this at once/

Reply:
If we put this operation within the lock we can only crack one analyzer assembly at a time. By keeping it outside of the lock we don't serialize them in the same way, at the risk of cracking the same assembly more than once.